cmd/soroban-rpc/internal/db/event.go

```diff
@@ -97,3 +109,169 @@ func (eventHandler *eventHandler) InsertEvents(lcm xdr.LedgerCloseMeta) error {
+// The events are returned in sorted ascending Cursor order.
+// If f returns false, the scan terminates early (f will not be applied on
+// remaining events in the range).
+func (eventHandler *eventHandler) GetEvents(
+	ctx context.Context,
+	cursorRange events.CursorRange,
+	contractIDs []string,
+	f ScanFunction,
+) error {
+	start := time.Now()
+
+	var rows []struct {
+		EventCursorID string              `db:"id"`
+		TxIndex       int                 `db:"application_order"`
+		Lcm           xdr.LedgerCloseMeta `db:"meta"`
+	}
+
+	rowQ := sq.
+		Select("e.id", "e.application_order", "lcm.meta").
+		From(eventTableName + " e").
+		Join(ledgerCloseMetaTableName + " lcm ON (e.ledger_sequence = lcm.sequence)").
+		Where(sq.GtOrEq{"e.id": cursorRange.Start.String()}).
+		Where(sq.Lt{"e.id": cursorRange.End.String()}).
+		OrderBy("e.id ASC")
+
+	if len(contractIDs) > 0 {
+		rowQ = rowQ.Where(sq.Eq{"e.contract_id": contractIDs})
+	}
+
+	if err := eventHandler.db.Select(ctx, &rows, rowQ); err != nil {
+		return fmt.Errorf(
+			"db read failed for start ledger cursor= %v contractIDs= %v: %w",
+			cursorRange.Start.String(),
+			contractIDs,
+			err)
+	} else if len(rows) < 1 {
+		return fmt.Errorf("no LCM found with requested event filters")
+	}
+
+	for _, row := range rows {
+		eventCursorID, txIndex, lcm := row.EventCursorID, row.TxIndex, row.Lcm
+		reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(eventHandler.passphrase, lcm)
+		if err != nil {
+			return fmt.Errorf("failed to index to tx %d in ledger %d: %w", txIndex, lcm.LedgerSequence(), err)
+		}
+
+		err = reader.Seek(txIndex - 1)
+		if err != nil {
+			return fmt.Errorf("failed to index to tx %d in ledger %d: %w", txIndex, lcm.LedgerSequence(), err)
+		}
+
+		ledgerCloseTime := lcm.LedgerCloseTime()
+		ledgerTx, err := reader.Read()
+		transactionHash := ledgerTx.Result.TransactionHash
+		diagEvents, diagErr := ledgerTx.GetDiagnosticEvents()
+
+		if diagErr != nil {
+			return fmt.Errorf("db read failed for Event Id %s: %w", eventCursorID, err)
+		}
+
+		// Find events based on filter passed in function f
+		for eventIndex, event := range diagEvents {
+			cur := events.Cursor{Ledger: lcm.LedgerSequence(), Tx: uint32(txIndex), Event: uint32(eventIndex)}
+			if f != nil && !f(event, cur, ledgerCloseTime, &transactionHash) {
+				return nil
+			}
+		}
+	}
+
+	eventHandler.log.
+		WithField("startLedgerSequence", cursorRange.Start.Ledger).
+		WithField("endLedgerSequence", cursorRange.End.Ledger).
+		WithField("duration", time.Since(start)).
+		Debugf("Fetched and decoded all the events with filters - contractIDs: %v ", contractIDs)
+
+	return nil
+}
+
+// GetLedgerRange returns the min/max ledger sequence numbers from the events table
+// TODO: Once we unify all the retention window we would keep only one GetLedgerRange logic
+
+func (eventHandler *eventHandler) GetLedgerRange(ctx context.Context) (ledgerbucketwindow.LedgerRange, error) {
```

Comment:
This can be simplified using min/max queries - https://github.com/stellar/soroban-rpc/pull/217/files#diff-0de90191c7a664162c6d9b9ddee3a00bf98a069c661fdaa3eb0bda6b728c22f7R143.

But yes, once these changes go in, we can just use 1 implementation for all 3 tables

Reply:
Yep, please simplify the query following the approach from https://github.com/stellar/soroban-rpc/pull/217

Reply:
Now that we are unifying all retention window to 1 day, It makes more sense to use the common `GetLedgerRange` function here. I will wait for Aditya's [change](https://github.com/stellar/soroban-rpc/pull/217/files#diff-0de90191c7a664162c6d9b9ddee3a00bf98a069c661fdaa3eb0bda6b728c22f7R143) to merge to main. otherwise it is gonna be a throwaway work on event side.